modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultConstantFoldingOptimizationPhase.java

```diff
@@ -824,6 +831,48 @@ public void visitInvokeCallMember(InvokeCallMemberNode irInvokeCallMemberNode, C
             int j = i;
             irInvokeCallMemberNode.getArgumentNodes().get(i).visit(this, (e) -> irInvokeCallMemberNode.getArgumentNodes().set(j, e));
         }
+        PainlessMethod method = irInvokeCallMemberNode.getDecorationValue(IRDMethod.class);
```

Comment:
This is the bit that runs the method at compile time if it has the annotation. We only implemented this for statically imported methods defined like `from_class` or defined with instance binding because that is all we needed. We did not implement this for regular methods or augments or `bound_to` declarations. All of those throw an error when you try to load the `PainlessLookup` to warn folks that we never implemented this feature for them.

Reply:
Added an issue that will allow annotations to determine if they're allowed on a certain whitelist type (https://github.com/elastic/elasticsearch/issues/68322).